fog/view/server/src/fog_view_router_admin_service.rs

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+use grpcio::{ChannelBuilder, RpcContext, RpcStatus, UnarySink};
+use itertools::Itertools;
+use mc_common::logger::{log, Logger};
+use mc_fog_api::{
+    view::AddShardRequest,
+    view_grpc::{FogViewRouterAdminApi, FogViewStoreApiClient},
+};
+use mc_fog_uri::FogViewStoreUri;
+use mc_util_grpc::{
+    rpc_invalid_arg_error, rpc_logger, rpc_precondition_error, send_result,
+    ConnectionUriGrpcioChannel, Empty,
+};
+use mc_util_metrics::SVC_COUNTERS;
+use std::{
+    collections::HashMap,
+    str::FromStr,
+    sync::{Arc, RwLock},
+};
+
+#[derive(Clone)]
+pub struct FogViewRouterAdminService {
+    shard_clients: Arc<RwLock<HashMap<FogViewStoreUri, Arc<FogViewStoreApiClient>>>>,
+    logger: Logger,
+}
+
+impl FogViewRouterAdminService {
+    pub fn new(
+        shard_clients: Arc<RwLock<HashMap<FogViewStoreUri, Arc<FogViewStoreApiClient>>>>,
+        logger: Logger,
+    ) -> Self {
+        Self {
+            shard_clients,
+            logger,
+        }
+    }
+
+    pub fn add_shard_impl(&mut self, shard_uri: &str, logger: &Logger) -> Result<Empty, RpcStatus> {
+        let view_store_uri = FogViewStoreUri::from_str(shard_uri).map_err(|_| {
+            rpc_invalid_arg_error(
+                "add_shard",
+                format!("Shard uri string {} is invalid", shard_uri),
+                logger,
+            )
+        })?;
+        let mut shard_clients = self.shard_clients.write().expect("RwLock Poisoned");
+        if shard_clients.keys().contains(&view_store_uri) {
+            let error = rpc_precondition_error(
+                "add_shard",
+                format!("Shard uri {} already exists in the shard list", shard_uri),
+                logger,
+            );
+            return Err(error);
+        }
```

Comment:
:question: What happens if a shard is reset?  Will we be stuck in an error state? 
Not sure if stale shard removal is coming in a later PR to address this possibility.

Reply:
What do you mean by reset? This API is only meant to add active shards to an already running FVR cluster. I imagine if a shard is "reset" (what I take to mean "deactivated") then we'd do so in a cluster upgrade rather than on the fly with an API.

Reply:
I'm thinking shard X goes down for some reason and comes back up trying to re-register itself, it will get an error saying it's already registered.
Maybe that's handled by clusters, but I'm pretty ignorant of how clusters work

Reply:
To clarify, this API is going to be called by a human operator when they want to add a new shard to the FVR's list of shards. It won't be called by a shard. 

In the failure case you're describing, Kubernetes, i.e the cluster manager, will try to reboot the shard and since the shard is already in the FVR's list of known shards it won't need to take additional action.